addon-test-support/@ember/test-helpers/wait-until.ts

```diff
@@ -8,6 +12,17 @@ type Falsy = false | 0 | '' | null | undefined;
 export interface Options {
   timeout?: number;
   timeoutMessage?: string;
+
+  /**
+   * Instrument `Ember.onerror` and reject, when it is called. This is useful
+   * for detecting that an operation inside a run loop has failed.
+   *
+   * This uses {@link setupOnerror}, so it will override any error listeners you
+   * might have set up before.
+   *
+   * @default true if the test context has been setup for usage with {@link setupOnerror}
+   */
+  rejectOnError?: boolean;
```

Comment:
I'm not really fond of this, because:

> it will override any error listeners you might have set up before.

> if the test context has been setup for usage with `setupOnerror`

This limitation arises from how `setupOnerror` works internally. Alternatively, we could _always_ instrument `Ember.onerror` in tests, somewhat like:

```ts
const originalOnError = Ember.onerror;
let onErrorOverride?: typeof Ember.onerror;

Ember.onerror = error => {
  // _Always_ let `waitUntil` know about the error.
  notifyInternalOnError(error);

  // If the user called `setupOnerror(...)`, call the override.
  if (onErrorOverride)
    return onErrorOverride(error);
  
  // Otherwise call the original `Ember.onerror`.
  return onErrorOverride?.(error)
};

export function setupOnerror(onError?: typeof Ember.onerror): void {
  let context = getContext();

  if (!context) {
    throw new Error('Must setup test context before calling setupOnerror');
  }

  if (!cachedOnerror.has(context)) {
    throw new Error(
      '_cacheOriginalOnerror must be called before setupOnerror. Normally, this will happen as part of your test harness.'
    );
  }

  if (typeof onError !== 'function') {
    onError = cachedOnerror.get(context);
  }

  onErrorOverride = onError;
}

export function resetOnerror(): void {
  let context = getContext();
  if (context && cachedOnerror.has(context)) {
    onErrorOverride = cachedOnerror.get(context);
  }
}
```

Reply:
Also, letting `rejectOnError` default to `true` will inevitably break a lot of existing test suites, that currently try to emulate this missing feature by using `setupOnerror` or manually hooking into `Ember.onerror`.

However, I believe that ultimately the behavior implemented in this PR is what a user would expect. So while this breaking change may be a pain, because it will require changing all tests dealing with such erros, I think that it's a _very_ welcome change.

_Maybe_ we could even conjur up a codemod to ease the transition. Or we could add a compatibility mode, that will essentially change the order of `notifyInternalOnError` and `onErrorOverride`, so that "old" tests just continue to work.

```ts
const originalOnError = Ember.onerror;
let onErrorOverride?: typeof Ember.onerror;
let preferOnErrorOverride = true;

Ember.onerror = error => {
  // Compatibility mode for old tests.
  if (preferOnErrorOverride && onErrorOverride)
    return onErrorOverride(error);

  // _Always_ let `waitUntil` know about the error.
  notifyInternalOnError(error);

  // If the user called `setupOnerror(...)`, call the override.
  if (onErrorOverride)
    return onErrorOverride(error);
  
  // Otherwise call the original `Ember.onerror`.
  return onErrorOverride?.(error)
};
```

We should also consider adding a setter trap for `Ember.onerror` to catch users manually overriding it in tests.